File: scala_code/src/main/scala/com/artvb/iq/chapter1/problem3/PermutationsEqual.scala
package com.artvb.iq.chapter1.problem3

import scala.collection.immutable.HashMap

/**
 * Created by artburkart on 10/25/14.
 */
object PermutationsEqual {
  def permutationsEqual(strA: String, strB: String): Boolean = {
    @annotation.tailrec
    def createCharMap(l: List[Char], m: HashMap[Char, Int] = new HashMap[Char, Int](), n: Int): HashMap[Char, Int] = l match {
      case (Nil) => m
      case (h :: t) => m.get(h) match {
        case (None) => if (n == 0) m else createCharMap(t, m + (h -> 1), n - 1)
        case Some(i) => if (n == 0 && i - 1 < 0) m else if (n == 0) createCharMap(t, m + (h -> (i - 1)), n) else createCharMap(t, m + (h -> (i + 1)), n - 1)
      }
    }
    if ((strA.length() != strB.length()))
      false
    else
      createCharMap((strA + strB).toList, new HashMap[Char, Int](), strA.length()).values.filter(_ != 0).toList.length == 0
  }
}

File: java_code/src/main/java/com/artvb/iq/chapter1/problem3/PermutationsEqual.java
package com.artvb.iq.chapter1.problem3;

import java.util.HashMap;

/**
 * Created by artburkart on 10/25/14.
 */
public class PermutationsEqual {
    /**
     * Checks if strA is a permutation of strB
     *
     * @param strA the first String
     * @param strB the second String
     * @return true iff strA is a permutation of strB, otherwise false
     */
    public static Boolean permutationsEqual(String strA, String strB) {
        if (strA == null || strB == null) {
            throw new NullPointerException();
        }
        if (strA.length() != strB.length()) {
            return false;
        }

        String curr;
        HashMap<String, Integer> tracker = new HashMap<String, Integer>();

        // Add 1 for every character in strA
        for (char c : strA.toCharArray()) {
            curr = String.valueOf(c);
            if (tracker.get(curr) == null) {
                tracker.put(curr, 1);
            } else {
                tracker.put(curr, tracker.get(curr) + 1);
            }
        }

        // Subtract 1 for every character in strB
        for (char c : strB.toCharArray()) {
            curr = String.valueOf(c);
            if (tracker.get(curr) == null) {
                return false;
            } else {
                tracker.put(curr, tracker.get(curr) - 1);
            }
            if (tracker.get(curr) < 0) {
                return false;
            }
        }

        for (int val : tracker.values()) {
           if (val > 0) {
               return false;
           }
        }
        return true;

    }
}
